freeipa/resource_freeipa_dns_record.go

```diff
@@ -0,0 +1,211 @@
+)
+
+func resourceFreeIPADNSRecord() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceFreeIPADNSRecordCreate,
+		Read:   resourceFreeIPADNSRecordRead,
+		Update: resourceFreeIPADNSRecordUpdate,
+		Delete: resourceFreeIPADNSRecordDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceFreeIPADNSRecordImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"idnsname": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"dnszoneidnsname": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"dnsttl": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"dnsclass": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"a_part_ip_address": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceFreeIPADNSRecordCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[INFO][freeipa] Creating DNS Record: %s", d.Id())
+
+	client, err := meta.(*Config).Client()
+	if err != nil {
+		return err
+	}
+
+	args := ipa.DnsrecordAddArgs{
+		Idnsname: d.Get("idnsname").(string),
+	}
+
+	optArgs := ipa.DnsrecordAddOptionalArgs{}
+
+	if _dnszoneidnsname, ok := d.GetOkExists("dnszoneidnsname"); ok {
+		dnszoneidnsname := _dnszoneidnsname.(string)
+		optArgs.Dnszoneidnsname = &dnszoneidnsname
+	}
+	
+	if _dnsttl, ok := d.GetOkExists("dnsttl"); ok {
+		dnsttl := _dnsttl.(int)
+		optArgs.Dnsttl = &dnsttl
+	}
+	
+	if _dnsclass, ok := d.GetOkExists("dnsclass"); ok {
+		dnsclass := _dnsclass.(string)
+		optArgs.Dnsclass = &dnsclass
+	}
+	
+	if _aPartIPAddress, ok := d.GetOkExists("a_part_ip_address"); ok {
+		aPartIPAddress := _aPartIPAddress.(string)
+		optArgs.APartIPAddress = &aPartIPAddress
+	}
+
+	_, err = client.DnsrecordAdd(&args, &optArgs)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(args.Idnsname)
+
+	return resourceFreeIPADNSRecordRead(d, meta)
+}
+
+func resourceFreeIPADNSRecordUpdate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[INFO] Updating DNS Record: %s", d.Id())
+
+	client, err := meta.(*Config).Client()
+	if err != nil {
+		return err
+	}
+
+	args := ipa.DnsrecordModArgs{
+		Idnsname: d.Get("idnsname").(string),
+	}
+
+	optArgs := ipa.DnsrecordModOptionalArgs{}
+
+	if _dnszoneidnsname, ok := d.GetOkExists("dnszoneidnsname"); ok {
+		dnszoneidnsname := _dnszoneidnsname.(string)
+		optArgs.Dnszoneidnsname = &dnszoneidnsname
+	}
+	
+	if _dnsttl, ok := d.GetOkExists("dnsttl"); ok {
+		dnsttl := _dnsttl.(int)
+		optArgs.Dnsttl = &dnsttl
+	}
+	
+	if _dnsclass, ok := d.GetOkExists("dnsclass"); ok {
+		dnsclass := _dnsclass.(string)
+		optArgs.Dnsclass = &dnsclass
+	}
+	
+	if _aPartIPAddress, ok := d.GetOkExists("a_part_ip_address"); ok {
+		aPartIPAddress := _aPartIPAddress.(string)
+		optArgs.APartIPAddress = &aPartIPAddress
+	}
+
+	_, err = client.DnsrecordMod(&args, &optArgs)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(args.Idnsname)
```

Comment:
I don't think there's a case where the ID would change during update. Is that really necessary?

Reply:
Removed